fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/DataCachePopulateRewriteRule.java

```diff
@@ -0,0 +1,184 @@
+        if (!sessionVariable.isEnablePopulateDataCache()) {
+            // be compatible with old parameter
+            populateMode = DataCachePopulateMode.NEVER;
+        }
+
+        // if populateMode is auto, we have to check further
+        if (populateMode == DataCachePopulateMode.AUTO) {
+            if (!isSatisfiedStatement()) {
+                populateMode = DataCachePopulateMode.NEVER;
+            }
+        }
+        return populateMode;
+    }
+
+    private boolean isSatisfiedStatement() {
+        // check is analyze sql
+        if (connectContext.isStatisticsJob()) {
+            return false;
+        }
+        // check is query statement
+        // make sure executor is not nullptr first
+        if (connectContext.getExecutor() == null) {
+            return false;
+        }
+        if (!(connectContext.getExecutor().getParsedStmt() instanceof QueryStatement)) {
+            return false;
+        }
+        return true;
+    }
+
+    private static class DataCachePopulateRewriteVisitor extends OptExpressionVisitor<Void, TaskContext> {
+        private final DataCachePopulateMode populateMode;
+
+        private DataCachePopulateRewriteVisitor(DataCachePopulateMode populateMode) {
+            this.populateMode = populateMode;
+        }
+
+        @Override
+        public Void visit(OptExpression optExpression, TaskContext context) {
+            for (OptExpression input : optExpression.getInputs()) {
+                input.getOp().accept(this, input, context);
+            }
+            return null;
+        }
+
+        @Override
+        public Void visitPhysicalScan(OptExpression optExpression, TaskContext context) {
+            PhysicalScanOperator scanOperator = (PhysicalScanOperator) optExpression.getOp();
+
+            switch (populateMode) {
+                case NEVER:
+                    return rewritePhysicalScanOperator(scanOperator, false);
+                case ALWAYS:
+                    return rewritePhysicalScanOperator(scanOperator, true);
+                default: {
+                    Table table = scanOperator.getTable();
+
+                    if (!isValidScanOperatorType(scanOperator.getOpType())) {
+                        return rewritePhysicalScanOperator(scanOperator, false);
+                    }
+
+                    // ignore full table scan
+                    if (checkIsFullColumnScan(table, scanOperator)) {
+                        return rewritePhysicalScanOperator(scanOperator, false);
+                    }
+
+                    ScanOperatorPredicates predicates = null;
+                    try {
+                        // ScanOperatorPredicates maybe nullptr, we have to check it
+                        predicates = scanOperator.getScanOperatorPredicates();
+                    } catch (AnalysisException e) {
+                        LOG.warn("Failed to get ScanOperatorPredicates", e);
+                    }
+
+                    if (predicates == null) {
```

Comment:
If select a single column for a unpartitioned table without predicates, will not populate cache?

Reply:
Actually `predicates` can't be nullptr, here just defense code.
Even a SQL without predicates, it will have an empty ScanOperatorPredicates and `predicates` can't be nullptr.